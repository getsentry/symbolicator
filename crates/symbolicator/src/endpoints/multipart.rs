//! Helper utilities to handle HTTP multipart bodies.

use std::io::SeekFrom;

use axum::extract::multipart::Field;
use axum::http::StatusCode;
use futures::prelude::*;
use tokio::fs::File;
use tokio::io::AsyncSeekExt;
use tokio_util::io::StreamReader;

use super::ResponseError;

/// Stream a multipart body to a file.
///
/// The file's cursor will be set to the beginning of the file again.
pub async fn stream_multipart_file(field: Field<'_>, file: &mut File) -> Result<(), ResponseError> {
    let mut field_reader =
        StreamReader::new(field.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err)));
    tokio::io::copy(&mut field_reader, file).await?;
    file.seek(SeekFrom::Start(0)).await?;
    Ok(())
}

/// Read a multipart body into memory.
///
/// This respects a maximum size.
pub async fn read_multipart_data(
    mut field: Field<'_>,
    max_size: usize,
) -> Result<Vec<u8>, ResponseError> {
    let mut data = Vec::with_capacity(512);

    while let Some(chunk) = field.next().await {
        let chunk = chunk?;
        if (data.len() + chunk.len()) > max_size {
            let err = anyhow::anyhow!(
                "Field {} larger than {} bytes",
                field.name().unwrap_or_default(),
                max_size
            );
            return Err((StatusCode::PAYLOAD_TOO_LARGE, err).into());
        }
        data.extend_from_slice(&chunk);
    }
    Ok(data)
}
