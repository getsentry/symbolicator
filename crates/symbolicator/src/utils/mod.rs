pub mod addr;
pub mod compression;
pub mod futures;
pub mod gcs;
pub mod hex;
pub mod http;
pub mod paths;
pub mod sentry;
