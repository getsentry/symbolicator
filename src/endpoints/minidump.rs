use actix_web::{
    dev::Payload, error, http::Method, multipart, Error, HttpMessage, HttpRequest, Json, Query,
    State,
};
use bytes::Bytes;
use futures::{FutureExt, TryFutureExt};
use futures01::{future, Future, Stream};
use sentry::{configure_scope, Hub};

use crate::actors::symbolication::SymbolicationActor;
use crate::app::{ServiceApp, ServiceState};
use crate::endpoints::symbolicate::SymbolicationRequestQueryParams;
use crate::sources::SourceConfig;
use crate::types::{RequestId, RequestOptions, Scope, SymbolicationResponse};
use crate::utils::futures::{ResponseFuture, ThreadPool};
use crate::utils::multipart::{
    read_multipart_file, read_multipart_request_options, read_multipart_sources,
};
use crate::utils::sentry::{ActixWebHubExt, SentryFutureExt, WriteSentryScope};

#[derive(Debug, Default)]
struct MinidumpRequest {
    sources: Option<Vec<SourceConfig>>,
    minidump: Option<Bytes>,
    options: RequestOptions,
}

fn handle_multipart_item(
    threadpool: ThreadPool,
    mut request: MinidumpRequest,
    item: multipart::MultipartItem<Payload>,
) -> ResponseFuture<MinidumpRequest, Error> {
    let field = match item {
        multipart::MultipartItem::Field(field) => field,
        multipart::MultipartItem::Nested(nested) => {
            return handle_multipart_stream(threadpool, request, nested);
        }
    };

    match field
        .content_disposition()
        .as_ref()
        .and_then(|d| d.get_name())
    {
        Some("sources") => {
            let future = read_multipart_sources(field).map(move |sources| {
                request.sources = Some(sources);
                request
            });
            Box::new(future)
        }
        Some("upload_file_minidump") => {
            let future = read_multipart_file(field).map(move |minidump| {
                request.minidump = Some(minidump);
                request
            });
            Box::new(future)
        }
        Some("options") => {
            let future = read_multipart_request_options(field).map(move |options| {
                request.options = options;
                request
            });
            Box::new(future)
        }
        _ => {
            // Always ignore unknown fields.
            Box::new(future::ok(request))
        }
    }
}

fn handle_multipart_stream(
    threadpool: ThreadPool,
    request: MinidumpRequest,
    stream: multipart::Multipart<Payload>,
) -> ResponseFuture<MinidumpRequest, Error> {
    let future = stream
        .map_err(Error::from)
        .fold(request, move |request, item| {
            handle_multipart_item(threadpool.clone(), request, item)
        });

    Box::new(future)
}

fn process_minidump(
    symbolication: &SymbolicationActor,
    request: MinidumpRequest,
    scope: Scope,
) -> Result<RequestId, Error> {
    let minidump = request
        .minidump
        .ok_or_else(|| error::ErrorBadRequest("missing minidump"))?;

    sentry::configure_scope(|scope| {
        let mut hasher = crc32fast::Hasher::new();
        hasher.update(&minidump);
        scope.set_extra("minidump_crc32", hasher.finalize().into());
        scope.set_extra("minidump_len", minidump.len().into());
    });

    let sources = request
        .sources
        .ok_or_else(|| error::ErrorBadRequest("missing sources"))?;

    Ok(symbolication.process_minidump(scope, minidump, sources, request.options))
}

fn handle_minidump_request(
    state: State<ServiceState>,
    params: Query<SymbolicationRequestQueryParams>,
    request: HttpRequest<ServiceState>,
) -> ResponseFuture<Json<SymbolicationResponse>, Error> {
    let hub = Hub::from_request(&request);
    hub.start_session();

    Hub::run(hub, || {
        let default_sources = state.config().default_sources();

        let params = params.into_inner();
        configure_scope(|scope| {
            params.write_sentry_scope(scope);
        });

        let request_future = handle_multipart_stream(
            state.io_pool(),
            MinidumpRequest::default(),
            request.multipart(),
        );

        let SymbolicationRequestQueryParams { scope, timeout } = params;
        let symbolication = state.symbolication();

        let response_future = request_future
            .and_then(clone!(symbolication, |mut request| {
                if request.sources.is_none() {
                    request.sources = Some(default_sources.to_vec());
                }

                process_minidump(&symbolication, request, scope)
            }))
            .and_then(move |request_id| {
                symbolication
                    .get_response(request_id, timeout)
                    .never_error()
                    .boxed_local()
                    .compat()
                    .then(|result| match result {
                        Ok(Some(response)) => Ok(Json(response)),
                        Ok(None) => Err(error::ErrorInternalServerError(
                            "symbolication request did not start",
                        )),
                        Err(never) => match never {},
                    })
                    .map_err(Error::from)
            });

        Box::new(response_future.sentry_hub_current())
    })
}

pub fn configure(app: ServiceApp) -> ServiceApp {
    app.resource("/minidump", |r| {
        r.method(Method::POST).with(handle_minidump_request);
    })
}

#[cfg(test)]
mod tests {
    use std::fs;

    use actix_web::test::TestServer;
    use reqwest::{multipart, Client, StatusCode};

    use crate::app::ServiceState;
    use crate::config::Config;
    use crate::test;
    use crate::types::SymbolicationResponse;

    #[tokio::test]
    async fn test_basic() {
        test::setup();

        let service = ServiceState::create(Config::default()).unwrap();
        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));

        let file_contents = fs::read("tests/fixtures/windows.dmp").unwrap();
        let file_part = multipart::Part::bytes(file_contents).file_name("windows.dmp");

        let form = multipart::Form::new()
            .part("upload_file_minidump", file_part)
            .text("sources", "[]");

        let response = Client::new()
            .post(&server.url("/minidump"))
            .multipart(form)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let body = response.text().await.unwrap();
        let response = serde_json::from_str::<SymbolicationResponse>(&body).unwrap();
        insta::assert_yaml_snapshot!(response);
    }

    // This test is disabled because it locks up on CI. We have not found a way to reproduce this.
    #[allow(dead_code)]
    // #[tokio::test]
    async fn test_integration_microsoft() {
        // TODO: Move this test to E2E tests
        test::setup();

        let service = ServiceState::create(Config::default()).unwrap();
        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));
        let source = test::microsoft_symsrv();

        let file_contents = fs::read("tests/fixtures/windows.dmp").unwrap();
        let file_part = multipart::Part::bytes(file_contents).file_name("windows.dmp");

        let form = multipart::Form::new()
            .part("upload_file_minidump", file_part)
            .text("sources", serde_json::to_string(&vec![source]).unwrap())
            .text("options", r#"{"dif_candidates":true}"#);

        let response = Client::new()
            .post(&server.url("/minidump"))
            .multipart(form)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let body = response.text().await.unwrap();
        let response = serde_json::from_str::<SymbolicationResponse>(&body).unwrap();
        insta::assert_yaml_snapshot!(response);
    }

    #[tokio::test]
    async fn test_unknown_field() {
        test::setup();

        let service = ServiceState::create(Config::default()).unwrap();
        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));

        let file_contents = fs::read("tests/fixtures/windows.dmp").unwrap();
        let file_part = multipart::Part::bytes(file_contents).file_name("windows.dmp");

        let form = multipart::Form::new()
            .part("upload_file_minidump", file_part)
            .text("sources", "[]")
            .text("unknown", "value");

        let response = Client::new()
            .post(&server.url("/minidump"))
            .multipart(form)
            .send()
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
    }
}
