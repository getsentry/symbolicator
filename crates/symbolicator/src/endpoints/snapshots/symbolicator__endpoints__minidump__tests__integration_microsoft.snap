---
source: src/endpoints/minidump.rs
expression: response
---
status: completed
timestamp: 1521713273
system_info:
  os_name: Windows
  os_version: 10.0.14393
  os_build: ""
  cpu_arch: x86
  device_model: ""
crashed: true
crash_reason: EXCEPTION_ACCESS_VIOLATION_WRITE
assertion: ""
stacktraces:
  - thread_id: 1636
    is_requesting: true
    registers:
      eax: "0x0"
      ebp: "0x10ff670"
      ebx: "0xfe5000"
      ecx: "0x10ff670"
      edi: "0x13bfd78"
      edx: "0x7"
      eflags: "0x10246"
      eip: "0x2a2a3d"
      esi: "0x759c6314"
      esp: "0x10ff644"
    frames:
      - status: missing
        original_index: 0
        instruction_addr: "0x2a2a3d"
        package: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
        trust: context
      - status: missing
        original_index: 1
        instruction_addr: "0x2a28d0"
        package: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
        trust: fp
      - status: symbolicated
        original_index: 2
        instruction_addr: "0x7584e9bf"
        package: "C:\\Windows\\System32\\rpcrt4.dll"
        symbol: "?FreeWrapper@@YGXPAX@Z"
        sym_addr: "0x7584e960"
        function: FreeWrapper(void *)
        lineno: 0
        trust: scan
      - status: symbolicated
        original_index: 3
        instruction_addr: "0x70b7ae3f"
        package: "C:\\Windows\\System32\\dbgcore.dll"
        symbol: "?DetermineOutputProvider@@YGJPAVMiniDumpAllocationProvider@@PAXQAU_MINIDUMP_CALLBACK_INFORMATION@@PAPAVMiniDumpOutputProvider@@@Z"
        sym_addr: "0x70b7ad6b"
        function: "DetermineOutputProvider(class MiniDumpAllocationProvider *,void *,struct _MINIDUMP_CALLBACK_INFORMATION * const,class MiniDumpOutputProvider * *)"
        lineno: 0
        trust: scan
      - status: symbolicated
        original_index: 4
        instruction_addr: "0x7584e9bf"
        package: "C:\\Windows\\System32\\rpcrt4.dll"
        symbol: "?FreeWrapper@@YGXPAX@Z"
        sym_addr: "0x7584e960"
        function: FreeWrapper(void *)
        lineno: 0
        trust: scan
      - status: missing
        original_index: 5
        instruction_addr: "0x2a3435"
        package: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
        trust: scan
      - status: missing
        original_index: 6
        instruction_addr: "0x2a2d97"
        package: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
        trust: scan
      - status: symbolicated
        original_index: 7
        instruction_addr: "0x750662c3"
        package: "C:\\Windows\\System32\\kernel32.dll"
        symbol: "@BaseThreadInitThunk@12"
        sym_addr: "0x750662a0"
        function: "@BaseThreadInitThunk@12"
        lineno: 0
        trust: fp
      - status: symbolicated
        original_index: 8
        instruction_addr: "0x771d0f78"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: __RtlUserThreadStart@8
        sym_addr: "0x771d0f4a"
        function: __RtlUserThreadStart@8
        lineno: 0
        trust: cfi
      - status: symbolicated
        original_index: 9
        instruction_addr: "0x771d0f43"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: _RtlUserThreadStart@8
        sym_addr: "0x771d0f29"
        function: _RtlUserThreadStart@8
        lineno: 0
        trust: cfi
  - thread_id: 3580
    is_requesting: false
    registers:
      eax: "0x0"
      ebp: "0x159faa4"
      ebx: "0x13b0990"
      ecx: "0x0"
      edi: "0x13b4af0"
      edx: "0x0"
      eflags: "0x216"
      eip: "0x771e016c"
      esi: "0x13b4930"
      esp: "0x159f900"
    frames:
      - status: symbolicated
        original_index: 0
        instruction_addr: "0x771e016c"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: ZwWaitForWorkViaWorkerFactory@20
        sym_addr: "0x771e0160"
        function: ZwWaitForWorkViaWorkerFactory@20
        lineno: 0
        trust: context
      - status: symbolicated
        original_index: 1
        instruction_addr: "0x771a6a10"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: TppWorkerThread@4
        sym_addr: "0x771a6770"
        function: TppWorkerThread@4
        lineno: 0
        trust: cfi
      - status: symbolicated
        original_index: 2
        instruction_addr: "0x750662c3"
        package: "C:\\Windows\\System32\\kernel32.dll"
        symbol: "@BaseThreadInitThunk@12"
        sym_addr: "0x750662a0"
        function: "@BaseThreadInitThunk@12"
        lineno: 0
        trust: cfi
      - status: symbolicated
        original_index: 3
        instruction_addr: "0x771d0f78"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: __RtlUserThreadStart@8
        sym_addr: "0x771d0f4a"
        function: __RtlUserThreadStart@8
        lineno: 0
        trust: cfi
      - status: symbolicated
        original_index: 4
        instruction_addr: "0x771d0f43"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: _RtlUserThreadStart@8
        sym_addr: "0x771d0f29"
        function: _RtlUserThreadStart@8
        lineno: 0
        trust: cfi
  - thread_id: 2600
    is_requesting: false
    registers:
      eax: "0x0"
      ebp: "0x169fb98"
      ebx: "0x13b0990"
      ecx: "0x0"
      edi: "0x13b7c28"
      edx: "0x0"
      eflags: "0x202"
      eip: "0x771e016c"
      esi: "0x13b7a68"
      esp: "0x169f9f4"
    frames:
      - status: symbolicated
        original_index: 0
        instruction_addr: "0x771e016c"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: ZwWaitForWorkViaWorkerFactory@20
        sym_addr: "0x771e0160"
        function: ZwWaitForWorkViaWorkerFactory@20
        lineno: 0
        trust: context
      - status: symbolicated
        original_index: 1
        instruction_addr: "0x771a6a10"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: TppWorkerThread@4
        sym_addr: "0x771a6770"
        function: TppWorkerThread@4
        lineno: 0
        trust: cfi
      - status: symbolicated
        original_index: 2
        instruction_addr: "0x750662c3"
        package: "C:\\Windows\\System32\\kernel32.dll"
        symbol: "@BaseThreadInitThunk@12"
        sym_addr: "0x750662a0"
        function: "@BaseThreadInitThunk@12"
        lineno: 0
        trust: cfi
      - status: symbolicated
        original_index: 3
        instruction_addr: "0x771d0f78"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: __RtlUserThreadStart@8
        sym_addr: "0x771d0f4a"
        function: __RtlUserThreadStart@8
        lineno: 0
        trust: cfi
      - status: symbolicated
        original_index: 4
        instruction_addr: "0x771d0f43"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: _RtlUserThreadStart@8
        sym_addr: "0x771d0f29"
        function: _RtlUserThreadStart@8
        lineno: 0
        trust: cfi
  - thread_id: 2920
    is_requesting: false
    registers:
      eax: "0x0"
      ebp: "0x179f2b8"
      ebx: "0x17b1aa0"
      ecx: "0x0"
      edi: "0x17b1a90"
      edx: "0x0"
      eflags: "0x206"
      eip: "0x771df3dc"
      esi: "0x2cc"
      esp: "0x179f2ac"
    frames:
      - status: symbolicated
        original_index: 0
        instruction_addr: "0x771df3dc"
        package: "C:\\Windows\\System32\\ntdll.dll"
        symbol: ZwGetContextThread@8
        sym_addr: "0x771df3d0"
        function: ZwGetContextThread@8
        lineno: 0
        trust: context
      - status: symbolicated
        original_index: 1
        instruction_addr: "0x76e75dbf"
        package: "C:\\Windows\\System32\\KERNELBASE.dll"
        symbol: NlsIsUserDefaultLocale@4
        sym_addr: "0x76e75d90"
        function: NlsIsUserDefaultLocale@4
        lineno: 0
        trust: cfi
modules:
  - debug_status: missing
    unwind_status: missing
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 5ab380779000
    code_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe"
    debug_id: 3249d99d-0c40-4931-8610-f4e4fb0b6936-1
    debug_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb"
    image_addr: "0x2a0000"
    image_size: 36864
    candidates:
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/crash.exe/5AB380779000/crash.ex_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/crash.exe/5AB380779000/crash.exe"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/crash.pdb/3249D99D0C4049318610F4E4FB0B69361/crash.pd_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/crash.pdb/3249D99D0C4049318610F4E4FB0B69361/crash.pdb"
        download:
          status: notfound
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: "57898e12145000"
    code_file: "C:\\Windows\\System32\\dbghelp.dll"
    debug_id: 9c2a902b-6fdf-40ad-8308-588a41d572a0-1
    debug_file: dbghelp.pdb
    image_addr: "0x70850000"
    image_size: 1331200
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 589abc846c000
    code_file: "C:\\Windows\\System32\\msvcp140.dll"
    debug_id: bf5257f7-8c26-43dd-9bb7-901625e1136a-1
    debug_file: msvcp140.i386.pdb
    image_addr: "0x709a0000"
    image_size: 442368
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 57898eeb92000
    code_file: "C:\\Windows\\System32\\apphelp.dll"
    debug_id: 8daf7773-372f-460a-af38-944e193f7e33-1
    debug_file: apphelp.pdb
    image_addr: "0x70a10000"
    image_size: 598016
  - debug_status: found
    unwind_status: found
    features:
      has_debug_info: true
      has_unwind_info: true
      has_symbols: true
      has_sources: false
    arch: x86
    type: pe
    code_id: 57898dab25000
    code_file: "C:\\Windows\\System32\\dbgcore.dll"
    debug_id: aec7ef2f-df4b-4642-a471-4c3e5fe8760a-1
    debug_file: dbgcore.pdb
    image_addr: "0x70b70000"
    image_size: 151552
    candidates:
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/dbgcore.dll/57898DAB25000/dbgcore.dl_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/dbgcore.dll/57898DAB25000/dbgcore.dll"
        download:
          status: ok
          features:
            has_debug_info: false
            has_unwind_info: false
            has_symbols: true
            has_sources: false
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/dbgcore.pdb/AEC7EF2FDF4B4642A4714C3E5FE8760A1/dbgcore.pd_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/dbgcore.pdb/AEC7EF2FDF4B4642A4714C3E5FE8760A1/dbgcore.pdb"
        download:
          status: ok
          features:
            has_debug_info: true
            has_unwind_info: true
            has_symbols: true
            has_sources: false
        unwind:
          status: ok
        debug:
          status: ok
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 589abc7714000
    code_file: "C:\\Windows\\System32\\VCRUNTIME140.dll"
    debug_id: 0ed80a50-ecda-472b-86a4-eb6c833f8e1b-1
    debug_file: vcruntime140.i386.pdb
    image_addr: "0x70c60000"
    image_size: 81920
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 57899141a000
    code_file: "C:\\Windows\\System32\\CRYPTBASE.dll"
    debug_id: 147c51fb-7ca1-408f-85b5-285f2ad6f9c5-1
    debug_file: cryptbase.pdb
    image_addr: "0x73ba0000"
    image_size: 40960
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 59bf30e31f000
    code_file: "C:\\Windows\\System32\\sspicli.dll"
    debug_id: 51e432b1-0450-4b19-8ed1-6d4335f9f543-1
    debug_file: wsspicli.pdb
    image_addr: "0x73bb0000"
    image_size: 126976
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 5a49bb7677000
    code_file: "C:\\Windows\\System32\\advapi32.dll"
    debug_id: 0c799483-b549-417d-8433-4331852031fe-1
    debug_file: advapi32.pdb
    image_addr: "0x73c70000"
    image_size: 487424
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 57899155be000
    code_file: "C:\\Windows\\System32\\msvcrt.dll"
    debug_id: 6f6409b3-d520-43c7-9b2f-62e00bfe761c-1
    debug_file: msvcrt.pdb
    image_addr: "0x73cf0000"
    image_size: 778240
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 598942c741000
    code_file: "C:\\Windows\\System32\\sechost.dll"
    debug_id: 6f6a05dd-0a80-478b-a419-9b88703bf75b-1
    debug_file: sechost.pdb
    image_addr: "0x74450000"
    image_size: 266240
  - debug_status: found
    unwind_status: found
    features:
      has_debug_info: true
      has_unwind_info: true
      has_symbols: true
      has_sources: false
    arch: x86
    type: pe
    code_id: 590285e9e0000
    code_file: "C:\\Windows\\System32\\kernel32.dll"
    debug_id: d3474559-96f7-47d6-bf43-c176b2171e68-1
    debug_file: wkernel32.pdb
    image_addr: "0x75050000"
    image_size: 917504
    candidates:
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/kernel32.dll/590285E9e0000/kernel32.dl_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/kernel32.dll/590285E9e0000/kernel32.dll"
        download:
          status: ok
          features:
            has_debug_info: false
            has_unwind_info: false
            has_symbols: true
            has_sources: false
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/wkernel32.pdb/D347455996F747D6BF43C176B2171E681/wkernel32.pd_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/wkernel32.pdb/D347455996F747D6BF43C176B2171E681/wkernel32.pdb"
        download:
          status: ok
          features:
            has_debug_info: true
            has_unwind_info: true
            has_symbols: true
            has_sources: false
        unwind:
          status: ok
        debug:
          status: ok
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 59b0df8f5a000
    code_file: "C:\\Windows\\System32\\bcryptPrimitives.dll"
    debug_id: 287b19c3-9209-4a2b-bb8f-bcc37f411b11-1
    debug_file: bcryptprimitives.pdb
    image_addr: "0x75130000"
    image_size: 368640
  - debug_status: found
    unwind_status: found
    features:
      has_debug_info: true
      has_unwind_info: true
      has_symbols: true
      has_sources: false
    arch: x86
    type: pe
    code_id: 5a49bb75c1000
    code_file: "C:\\Windows\\System32\\rpcrt4.dll"
    debug_id: ae131c67-27a7-4fa1-9916-b5a4aef41190-1
    debug_file: wrpcrt4.pdb
    image_addr: "0x75810000"
    image_size: 790528
    candidates:
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/rpcrt4.dll/5A49BB75c1000/rpcrt4.dl_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/rpcrt4.dll/5A49BB75c1000/rpcrt4.dll"
        download:
          status: ok
          features:
            has_debug_info: false
            has_unwind_info: false
            has_symbols: true
            has_sources: false
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/wrpcrt4.pdb/AE131C6727A74FA19916B5A4AEF411901/wrpcrt4.pd_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/wrpcrt4.pdb/AE131C6727A74FA19916B5A4AEF411901/wrpcrt4.pdb"
        download:
          status: ok
          features:
            has_debug_info: true
            has_unwind_info: true
            has_symbols: true
            has_sources: false
        unwind:
          status: ok
        debug:
          status: ok
  - debug_status: unused
    unwind_status: unused
    features:
      has_debug_info: false
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe
    code_id: 59bf2b5ae0000
    code_file: "C:\\Windows\\System32\\ucrtbase.dll"
    debug_id: 6bedcbce-0a3a-40e9-8040-81c2c8c6cc2f-1
    debug_file: ucrtbase.pdb
    image_addr: "0x758f0000"
    image_size: 917504
  - debug_status: found
    unwind_status: unused
    features:
      has_debug_info: true
      has_unwind_info: true
      has_symbols: true
      has_sources: false
    arch: x86
    type: pe
    code_id: 59bf2bcf1a1000
    code_file: "C:\\Windows\\System32\\KERNELBASE.dll"
    debug_id: 8462294a-c645-402d-ac82-a4e95f61ddf9-1
    debug_file: wkernelbase.pdb
    image_addr: "0x76db0000"
    image_size: 1708032
    candidates:
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/KERNELBASE.dll/59BF2BCF1a1000/KERNELBASE.dl_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/KERNELBASE.dll/59BF2BCF1a1000/KERNELBASE.dll"
        download:
          status: ok
          features:
            has_debug_info: false
            has_unwind_info: false
            has_symbols: true
            has_sources: false
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/wkernelbase.pdb/8462294AC645402DAC82A4E95F61DDF91/wkernelbase.pd_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/wkernelbase.pdb/8462294AC645402DAC82A4E95F61DDF91/wkernelbase.pdb"
        download:
          status: ok
          features:
            has_debug_info: true
            has_unwind_info: true
            has_symbols: true
            has_sources: false
        debug:
          status: ok
  - debug_status: found
    unwind_status: found
    features:
      has_debug_info: true
      has_unwind_info: true
      has_symbols: true
      has_sources: false
    arch: x86
    type: pe
    code_id: 59b0d8f3183000
    code_file: "C:\\Windows\\System32\\ntdll.dll"
    debug_id: 971f98e5-ce60-41ff-b2d7-235bbeb34578-1
    debug_file: wntdll.pdb
    image_addr: "0x77170000"
    image_size: 1585152
    candidates:
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/ntdll.dll/59B0D8F3183000/ntdll.dl_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/ntdll.dll/59B0D8F3183000/ntdll.dll"
        download:
          status: ok
          features:
            has_debug_info: false
            has_unwind_info: false
            has_symbols: true
            has_sources: false
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/wntdll.pdb/971F98E5CE6041FFB2D7235BBEB345781/wntdll.pd_"
        download:
          status: notfound
      - source: microsoft
        location: "https://msdl.microsoft.com/download/symbols/wntdll.pdb/971F98E5CE6041FFB2D7235BBEB345781/wntdll.pdb"
        download:
          status: ok
          features:
            has_debug_info: true
            has_unwind_info: true
            has_symbols: true
            has_sources: false
        unwind:
          status: ok
        debug:
          status: ok
